04/app.js

```diff
@@ -0,0 +1,39 @@
+const divList = document.querySelectorAll("div");
+const body = document.querySelector("body");
+
+divList.forEach(colorSquares);
+
+body.addEventListener("click", function (e) {
+  if (e.target !== this) {
+    return null;
+  }
+  divList.forEach(function (element) {
+    element.classList.remove("clicked");
+  });
+});
+
+function colorSquares(element) {
+  if (element.getAttribute("class") === "child") {
+    element.addEventListener("click", ColorAllSquares);
+  } else {
+    element.addEventListener("click", ColorTheSquare);
+  }
+}
+
+function ColorAllSquares(e) {
+  e.stopPropagation();
+  divList.forEach(function (ele) {
+    const timer = ele.dataset.time;
+    setTimeout(() => {
+      ele.classList.add("clicked");
+    }, timer);
+  });
+}
+
+function ColorTheSquare(e) {
+  e.stopPropagation();
+  const timer = this.dataset.time;
+  setTimeout(() => {
+    this.classList.add("clicked");
+  }, timer);
+}
```

Comment:
Wystarczy skorzystać z propagacji i nic więcej nam nie trzeba :)

```
body.addEventListener("click", function (e) {
  if (e.target !== this) {
    return null;
  }
  divList.forEach(function (element) {
    element.classList.remove("clicked");
  });
});

function colorSquares(element) {
  // if (element.getAttribute("class") === "child") {
  //   element.addEventListener("click", ColorAllSquares);
  // } else {
    element.addEventListener("click", ColorTheSquare);
  // }
}

// function ColorAllSquares(e) {
//   e.stopPropagation();
//   divList.forEach(function (ele) {
//     const timer = ele.dataset.time;
//     setTimeout(() => {
//       ele.classList.add("clicked");
//     }, timer);
//   });
// }

function ColorTheSquare(e) {
  // e.stopPropagation();
  const timer = this.dataset.time;
  setTimeout(() => {
    this.classList.add("clicked");
  }, timer);
}
```

PS. Nazwy funkcji piszemy  z wielkiej litery tylko wtedy kiedy to konstruktor (lub komponent React) - dlatego tutaj nie powinniśmy pisać ich z wielkiej litery